test/testdrive/kafka-avro-debezium-transaction.td

```diff
@@ -0,0 +1,365 @@
+
+# Note that this should still work even if data2 (which shares the tx_metadata topic) isn't able to progress!
+> SELECT a, b FROM data_schema_inline
+a  b
+-----
+1  1
+2  3
+4  5
+
+$ kafka-ingest format=avro topic=data-txdata schema=${txschema} timestamp=2
+{"status": "BEGIN", "id": "5", "event_count": null, "data_collections": null}
+{"status": "END", "id": "5", "event_count": {"long": 1}, "data_collections": {"array": [{"event_count": 1, "data_collection": "testdrive-data-${testdrive.seed}"}]}}
+
+
+$ kafka-ingest format=avro topic=data schema=${schema} timestamp=1
+{"before": null, "after": {"row": {"a": 8, "b": 9}}, "source": {"file": "binlog2", "pos": 1, "row": 1, "snapshot": {"boolean": false}}, "transaction": {"total_order": null, "id": "5"}}
+
+
+> CREATE SINK data_sink FROM data_schema_inline
+  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-data-sink-${testdrive.seed}'
+  WITH (reuse_topic=true) FORMAT AVRO
+  USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}'
+
+# Check that repeated Debezium messages are skipped.
+$ kafka-ingest format=avro topic=data schema=${schema} timestamp=1
+{"before": null, "after": {"row": {"a": 4, "b": 5}}, "source": {"file": "binlog", "pos": 1, "row": 1, "snapshot": {"boolean": false}}, "transaction": {"total_order": null, "id": "1"}}
+{"before": null, "after": {"row": {"a": 8, "b": 9}}, "source": {"file": "binlog2", "pos": 1, "row": 1, "snapshot": {"boolean": false}}, "transaction": {"total_order": null, "id": "5"}}
+
+> SELECT a, b FROM data_schema_inline
+a  b
+----
+1  1
+2  3
+4  5
+8  9
+
+# Now do data2
+$ kafka-ingest format=avro topic=data2 schema=${schema} timestamp=1
+{"before": null, "after": {"row": {"a": 101, "b": 101}}, "source": {"file": "binlog", "pos": 0, "row": 0, "snapshot": {"boolean": false}}, "transaction": {"total_order": null, "id": "1"}}
+
+> SELECT a, b FROM data2_schema_inline
+a    b
+---------
+101  101
+
+
+#
+# Create a source using a file schema
+#
+! CREATE MATERIALIZED SOURCE data_schema_file
+  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-data-${testdrive.seed}'
+  FORMAT AVRO USING SCHEMA FILE 'data-schema.json'
+  ENVELOPE DEBEZIUM
+contains:No such file or directory
+
+$ file-append path=data-schema.json
+\${schema}
+
+> CREATE MATERIALIZED SOURCE data_schema_file
+  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-data-${testdrive.seed}'
+  WITH(tx_metadata = data_txdata, tx_metadata_collection_name = 'testdrive-data-${testdrive.seed}')
+  FORMAT AVRO USING SCHEMA FILE '${testdrive.temp-dir}/data-schema.json'
+  ENVELOPE DEBEZIUM
+
+> SELECT a, b FROM data_schema_file
+a  b
+----
+1  1
+2  3
+4  5
+8  9
+
+$ set-regex match=\d{13} replacement=<TIMESTAMP>
+
+$ kafka-verify format=avro sink=materialize.public.data_sink sort-messages=true
```

Comment:
So one of the things that we haven't tested but I think it's worth adding is that the transaction boundaries are maintained and reflected by our timestamp assignment. It's a bit hard to do entirely within materialize because afaik the only command that exposes the underlying timestamps is TAIL which cannot be used in more complex queries and we don't really know the assigned timestamp to predict the output.

One thing we can do however is create a kafka source pointed to the data_sink that mz produces and then for example run a `SELECT COUNT(*) AS event_count GROUP BY transaction.id ORDER BY transaction.id` query that will verify that there were 3 distinct transactions with the expected number of transactions.

Reply:
clever!  will update